yt_dlp/extractor/common.py

```diff
@@ -1428,6 +1428,39 @@ def extract_interaction_statistic(e):
                     continue
                 info[count_key] = interaction_count
 
+        def extract_chapter_information(e):
+            chapters = []
+            last_chapter = {}
+            if not e.get('hasPart'):
+                return
+            for idx, part in enumerate(e['hasPart']):
+                if part.get('@type') != 'Clip':  # https://schema.org/Clip
+                    continue
+                if idx == 0:
+                    last_chapter = {
+                        'start_time': part.get('startOffset') or 0,
+                        'end_time': part.get('endOffset'),
+                    }
+                    last_chapter.update({'title': part['name']} if part.get('name') else {})
+                else:
+                    # Fix up the previous chapter using this chapter's data
+                    last_chapter['end_time'] = last_chapter.get('end_time') or part.get('startOffset')
+                    if None in last_chapter.values():
+                        # If there's null values now, the chapter listing is broken somehow. Emit warning and bail.
+                        self.report_warning(f'Chapter {last_chapter.get("title") or idx} contains broken data. Not extracting chapters.')
+                        return
+                    else:
+                        chapters.append(last_chapter)
+                    last_chapter = {
+                        'start_time': part.get('startOffset') or last_chapter.get('end_time'),
+                        'end_time': part.get('endOffset')
+                    }
+                    last_chapter.update({'title': part['name']} if part.get('name') else {})
+            if chapters:
+                last_chapter.update({} if last_chapter.get('end_time') else {'end_time': info['duration']})
+                chapters.append(last_chapter)
+                info['chapters'] = chapters
+
```

Comment:
why is this so complicated?
(untested)
```py
        def extract_chapter_information(e):
            chapters = [{
                'title': part.get('name'),
                'start_time': part.get('startOffset'),
                'end_time': part.get('endOffset'),
            } for part in (e.get('hasPart') or []) if part.get('@type') != 'Clip']
            if chapters and chapters[0]['start_time']:  # Chapters must always at 0
                chapters[:0] = [{'start_time': 0, 'end_time': None, 'title': '<Untitled>'}]
            for idx, (last_c, current_c, next_c) in enumerate(zip(
                    [{'end_time': 0}] + chapters, chapters, chapters[1:])):
                current_c['end_time'] = current_c['end_time'] or next_c['start_time']
                current_c['start_time'] = current_c['start_time'] or last_c['end_time']
                if None in current_c.values():
                    self.report_warning(f'Chapter {idx} contains broken data. Not extracting chapters')
                    return
            info['chapters'] = chapters
```

Reply:
I made it more verbose so I could debug it more easily while building it. Will look at your solution the coming days.

I don't quite get the "chapters must always start at 0" thing: the [`InfoExtractor` docs](https://github.com/yt-dlp/yt-dlp/blob/e978789f0f0bfe1963f4a295f6094dafa7524933/yt_dlp/extractor/common.py#L313) don't mention this, and `--embed-chapters` worked fine. Does something else break if they don't?

Reply:
While `--embed-chapters` work, the start of the first chapter gets automatically moved to 0 secs by ffmpeg. I think it is more appropriate to add a separate chapter in the start instead

Reply:
Huh, interesting. ffmpeg seems to *write* the metadata readably as VLC will correctly pick up the chapters (using its own, non-FFmpeg parser):
![VLC screencap](https://cloud.teutsch.me/s/iiEenrAF2ELXqyb/download/3nhO4rLsWc.gif)
Looks like something goes wrong while reading the chapters from the file - even on the current `master` branch. Anyway, that's not an issue we can actually fix here, so I guess the empty-chapter workaround will work for now...

Reply:
Revisiting this, isn't the empty-first-chapter thing something that should rather go into one of the post-extraction stages? This is an ffmpeg-specific issue and once it gets fixed, we would have to revisit every extractor that handles chapters and remove the generated empty chapter

Reply:
You dont need to add the empty chapter. I was not aware it is allowed. Since ffmpeg is writing it correctly, we dont need to do anything about it